### PR TITLE
Verify member expressions [blocks: #4023]

### PR DESCRIPTION
**Pull request by smowton**
Depends on #3966

This checks that the component they refer to really exists, and the expression and component's types match.

### CODE REVIEW

File: src/util/std_expr.h [hunk 1]
```diff
@@ -3933,7 +3948,6 @@ inline void validate_expr(const member_exprt &value)
   validate_operands(value, 1, "Extract member must have one operand");
 }
 
-
```
Comment (tautschnig) [inline src/util/std_expr.h]: Nit pick: unrelated change?

**Review by tautschnig — APPROVED** [review]

**Comment by kroening** [comment]
Please no `base_type_eq`?

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5dd83b10d22cbc7b0240d3255454953fe7ca5e6b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99665060
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by smowton** [comment]
No can do, lots of stuff fails if we have both `==` and `--validate-ssa-equation`, I think because of type renaming. With `base_type_eq` everything passes but generics will be ignored, because the follow operation skips them.

So which do you prefer:
1. Eliminate base_type_eq (find and fix all the follows)
2. Augment base_type_eq a bit to verify that when both types are tags, their decorations match (i.e. like my pointer-decoration PR)

Advantage of 1: clean
Advantage of 2: can do immediately and get some soundness checking in the meantime.

**Comment by smowton** [comment]
Aha, this can be done once limited-type-renaming is in -- by coincidence, in Java types that require renaming (only VLA data allocations) are mutually exclusive with those that use decorated tags, and therefore lose information on type following. Therefore this now works if we first get #3966 and #4126.

Given this narrow tightrope re: type renaming vs. decorated tags, I'll also make a PR to assert that renaming doesn't discard information.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7fb60d712cc0fcba567a1bc47c0147a02caf23a5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100218684

**Comment by smowton** [comment]
I expected this would be good now, but it isn't because there are cases where users of `base_type_eq` conflate qualified and unqualified generic types, especially symex-dereference. Therefore this needs to wait for #4056.

File: src/util/std_expr.cpp [hunk 2]
```diff
@@ -223,3 +224,34 @@ const exprt &object_descriptor_exprt::root_object() const
 
   return *p;
 }
+
+void member_exprt::validate(
+  const exprt &expr,
+  const namespacet &ns,
+  const validation_modet vm)
```
Comment (romainbrenguier) [inline src/util/std_expr.cpp:241]: Avoid abbreviations

File: src/util/std_expr.cpp [hunk 3]
```diff
@@ -223,3 +224,34 @@ const exprt &object_descriptor_exprt::root_object() const
...
+{
+  check(expr, vm);
+
+  const auto &member_expr = to_member_expr(expr);
```
Comment (romainbrenguier) [inline src/util/std_expr.cpp:245]: :question: why not take a `member_exprt` as argument instead, and let the caller call `to_member_expr` if needed?

File: src/util/std_expr.cpp [hunk 4]
```diff
@@ -223,3 +224,34 @@ const exprt &object_descriptor_exprt::root_object() const
...
+
+  const typet &compound_type = ns.follow(member_expr.compound().type());
+  DATA_CHECK(
+    vm,
+    can_cast_type<struct_union_typet>(compound_type),
+    "member must address a struct, union or compatible type");
+
+  const auto &component = to_struct_union_type(compound_type)
```
Comment (romainbrenguier) [inline src/util/std_expr.cpp]: :question: can't we use a `try_cast_dynamic_type` instead of `can_cast_type` + `to_struct_union_type`?

File: src/util/std_expr.h [hunk 5]
```diff
@@ -3897,6 +3897,21 @@ class member_exprt:public unary_exprt
   {
     return op0();
   }
+
+  static void check(
```
Comment (romainbrenguier) [inline src/util/std_expr.h:3799]: The two functions you introduce should be documented

**Comment by tautschnig** [comment]
@smowton This needs a rebase and addressing @romainbrenguier's comments, then it's ready to go in.

File: src/util/std_expr.cpp (hunk 2 above)
Comment (smowton) [reply src/util/std_expr.cpp:241]: Every other `validate` method is doing the same thing, so I'll leave this as-is for now (change them en masse if we want to)

File: src/util/std_expr.cpp (hunk 3 above)
Comment (smowton) [reply src/util/std_expr.cpp:245]: Appears to be normal for `validate` methods, again change en masse or not at all

File: src/util/std_expr.h (hunk 5 above)
Comment (tautschnig) [reply src/util/std_expr.h:3799]: @smowton Still to be done?

**Comment by smowton** [comment]
@tautschnig done; @romainbrenguier addressed comments; have also tested current test-gen against develop+this PR and found tests passing.

File: src/util/std_expr.h (hunk 5 above)
Comment (smowton) [reply src/util/std_expr.h:3799]: Is there really any value in documenting one-liner that says "assert(ops.count == 1)"?

Comment (tautschnig) [reply src/util/std_expr.h:3799]: I'm not so sure, I was just confused by you giving a thumbs-up but then seemingly no such change was included.

Comment (smowton) [reply src/util/std_expr.h:3799]: I documented `validate`, which is much more substantial

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 44756ee4a4dbda9a583339a1dec786c4160a7f3f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103522773
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.
